articles/generalized_interface_generation.md

```diff
@@ -0,0 +1,668 @@
+
+### Build system integration
+
+It can be achieved by adding thin layers that leverage [build system configuration](#build-system-configuration) and [meta build](#meta-build-tool) tooling to bootstrap an package builds.
+
+For instance, an `ament_cmake` integration would generate and include CMake code during the configuration stage e.g.:
+
+```cmake
+# Generate C++ type representation build specification
+execute_process(
+  COMMAND rosidl_build_configure
+     --language cpp
+     -I path/to/interface/dependencies/
+     interface0.idl interface1.idl
+  OUTPUT_FILE build.spec)
+# Translate build specification into CMake code
+execute_process(
+  COMMAND ament_meta_build -b cmake
+  INPUT_FILE build.spec
+  OUTPUT_FILE build.cmake)
+# Use generate CMake code
+include(build.cmake)
+```
+
+It is worth noting that these integration layers must bring in package dependencies on their own, as there is no build system-agnostic mechanism to do so.
+
+### Migration
+
+Migrating the current architecture is fairly straightforward.
+
+- Each tightly coupled pair of plugins for `rosidl_generate` and `rosidl_build_configure` can be built on top of each existing generator or type support package.
+- Generated build specifications can delegate to CMake while the `ament_meta_build` tooling is independently developed.
+- New `ament_cmake` macros can be added to tap into the pipeline, while keeping existing ones (which may be deprecated over time, or re-implemented on top of the former). 
+
+## Appendix
+
+### A. Common Build Specification
+
+The Common Build Specification is a mechanism for describing how to build and distribute a software package, such that any build tool can carry out the process. The draft specification can be found [here](https://github.com/hidmic/cps).
+
+This specification is based on the [Common Package Specification](https://mwoehlke.github.io/cps/) format, a simple and relatively standard format that describes packages for downstream consumption in a build tool-agnostic way.
+The CPS format was not designed to describe builds, and thus why a modification is needed.
+
+As examples, see below the `.cbs` files that would result from running `rosidl_generator_c` and `rosidl_generator_py` on a `foo_msgs` package.
+
+
+*foo_msgs_c.cbs*
+```json
+{
+    "Name": "foo_msgs_c",
+    "Description": "Build specification for ROS2 C foo_msgs interfaces",
+    "License": "Apache License 2",
+    "Version": "0.1.0",
+    "Compat-Version": "0.8.0",
+    "Platforms": {
+        "GNU" : {
+            "Cpp-Compiler": "gnu",
+        },
+        "Clang" : {
+            "Cpp-Compiler": "clang",
+        },
+        "Windows": {
+            "Kernel-Version": "windows"
+        }
+    },
+    "Requires": {
+        "rosidl_runtime_c": {},
+        "rosidl_typesupport_interface": {},
+        "action_msgs": {}
+    },
+    "Configurations": [ "Optimized", "Debug" ],
+    "Default-Components": [ "c" ],
+    "Components": {
+        "idls": {
```

Comment:
> Are the component names arbitrary?

They are, though consistency is required to establish dependencies between them (e.g. to build Python generated code, C generated code may be necessary). 
There's also something to be said about components shared by multiple build configurations. It is likely that components like `idls` will be required by many source code generators and since each build configuration is carried out independently, replication will ensue. By keeping names consistent, deduplication  becomes feasible when multiple generators are configured at once (e.g. when targeting all available generators by not specifying any on `rosidl build-configure`).

> What if there are no `.msg/.srv/.action` files to convert? Will `.idls` still exist and be a noop, or does that mean there will be no `idls` component?

Either would work. During build configuration, the tool has access to sources, so it can adapt the resulting `.cbs` accordingly (e.g. adding or removing `c_code` dependencies).

Reply:
> assuming the order that build-configure plugins are invoked is random

One note about this. Some topological ordering will necessarily happen somewhere: in `rosidl build-configure` by ordering plugins or when merging build configurations, in the build system after build configurations have been transformed, in the build tool if the user splits build configuration in multiple packages. I'm inclined to do so upon merging build configurations, simply to make things easier for the build system.

Reply:
>> What if there are no .msg/.srv/.action files to convert? Will .idls still exist and be a noop, or does that mean there will be no idls component?
>
> Either would work. During build specification, the tool has access to sources, so it can adapt the resulting .cbs accordingly (e.g. adding or removing c_code dependencies).

> By keeping names consistent, deduplication becomes feasible when multiple generators are configured at once (e.g. when targeting all available generators by not specifying any on rosidl build-configure).

So if all the `build-configure` plugins  know their respective `generator` plugins require `.idl` files but only `.msg` files are available, then the `build-configure` plugins all depend on `:idls` and add an `idls` component to translate any `.msg/.srv/.action` files to `.idl` files? Then `rosidl build-configure` de-duplicates by keeping only one `:idls` components?  If so, it seems like that requires all `build-configure` plugins to have exactly the same idea about how to convert files. If something changes and a plugin in the wild doesn't get updated, then all would be affected if `rosidl build-configure` deduplicates and keeps the out-of date `:idls` component.

Reply:
> So if all the `build-configure` plugins know their respective generator plugins require `.idl` files but only `.msg` files are available, then the build-configure plugins all depend on `:idls` and add an `:idls` component to translate any `.msg/.srv/.action` files to `.idl` files? 

Correct.

> Then `rosidl build-configure` de-duplicates by keeping only one `:idls` components?

Correct, assuming these components are all equivalent.

> it seems like that requires all build-configure plugins to have exactly the same idea about how to convert files.

That's right, if de-duplication is naive and only looks at component names. 

IIRC this design document does not describe a build specification merge procedure. I didn't get that far. Thinking out loud, to safely de-duplicate two or more components, these have to be equal. Equality here means that name and description are equal, as well as that of its local dependencies, recursively. 

For instance, say we have two `:foo` components:

```sh
":foo": {
     "Requires":  {":bar", "fizz:buzz"},
     ...
}
```

These can be deemed equal if their descriptions match. But that's not enough, since each could be using a different `:bar` (note that's not a problem for `fizz:buzz`, as it is a fully qualified component in an external package[1]), so those have to be equal too. And the same goes for `:bar` dependencies, and the dependencies of its dependencies, and so on. In practice, I highly doubt we'll have seemingly equivalent components like these.

[1] One could argue that two build specifications could depend on different versions of the same external package. The simplest solution is to ban merging them. If there's a need, the procedure could try to pick a version that satisfies both constraints, but it seems overkill to me.